### PR TITLE
[MSHARED-1080] Parent POM 36, Java8, drop legacy.

### PR DESCRIPTION
**Pull request by cstamas**
Changes:
* parent POM 36
* Java8, up locked deps (due Java7)
* drop maven-shared-util and other dead dependencies

---

https://issues.apache.org/jira/browse/MSHARED-1080

### CODE REVIEW

File: src/main/java/org/apache/maven/shared/filtering/DefaultMavenFileFilter.java [hunk 1]
```diff
@@ -105,8 +110,6 @@ public void copyFile( File from, File to, boolean filtering, List<FileUtils.Filt
                 }
                 FileUtils.copyFile( from, to, encoding, new FileUtils.FilterWrapper[0], overwrite );
             }
-
-            buildContext.refresh( to );
```
Comment (olamy) [inline src/main/java/org/apache/maven/shared/filtering/DefaultMavenFileFilter.java:109]: isn't something used by m2e? or not anymore?

Comment (cstamas) [reply src/main/java/org/apache/maven/shared/filtering/DefaultMavenFileFilter.java:109]: This comes from a repo that is last touched 12 years ago, and is archived. IMHO is dead, parrot dead, but we can tinker about it if you want :smile:

Comment (olamy) [reply src/main/java/org/apache/maven/shared/filtering/DefaultMavenFileFilter.java:109]: a quick search on github shows this https://github.com/search?q=org%3Aeclipse-m2e+BuildContext&type=code
for sure this very old repo is just interface which doesn't need much changes so that's probably the reason.

disclaimer i'm not using m2e so... 🤣

Comment (cstamas) [reply src/main/java/org/apache/maven/shared/filtering/DefaultMavenFileFilter.java:109]: You are missing the point: this parrot is dead. Very. Also, whatever it is, this was first attempt to implement incremental build, followed by two other attempts (by same person), that finally materialized in something completely different. So, yes, this may look "stable", the very same way dead parrot looks "stable" (well, not moving, is it?), but is superseded by several other attempts...

Comment (olamy) [reply src/main/java/org/apache/maven/shared/filtering/DefaultMavenFileFilter.java:109]: maybe but I can see references of it in the m2e code.
Is it really used or not? I have no idea as I'm not a contributor of m2e.
My opinion is just it worth to ask those folks as we may or not break something.
if not perfect but it worth the courtesy of simply asking.

**Comment by olamy** [comment]
@fbricon still active m2e? any idea if this BuildContext is still in use? or maybe you can ping someone else? thanks

File: src/main/java/org/apache/maven/shared/filtering/DefaultMavenFileFilter.java (hunk 1 above)
Comment (cstamas) [reply src/main/java/org/apache/maven/shared/filtering/DefaultMavenFileFilter.java:109]: Frankly, this should be other way around, don't you agree? This dead parrot (not authored by us, not used by anything other than m2e) does not buy anything for Maven itself (just adds extra complexity and more dependencies), does not buy anything for any other IDE out there either and finally does not buy anything for end users either. There is one and only one "beneficiary" of this code, if any.

In other words, if tomorrow some Big Team approaches us, wanting to put some "custom code" into Maven Core or Plugins to make it work for them, what would you say? And next day another team, and so on... 

Anyway, we are trying to move forward, spring cleanup time (pending since a lot years), and we do accept quality patches as well, so nothing is lost here IMO by removing this.

Comment (olamy) [reply src/main/java/org/apache/maven/shared/filtering/DefaultMavenFileFilter.java:109]: fyi there are still some interest here https://github.com/codehaus-plexus/plexus-build-api/issues

**Comment by olamy** [comment]
@rakus ping

File: src/main/java/org/apache/maven/shared/filtering/DefaultMavenFileFilter.java (hunk 1 above)
Comment (cstamas) [reply src/main/java/org/apache/maven/shared/filtering/DefaultMavenFileFilter.java:109]: heh, so yet another project that is superseded but is revived and babysit? Well, the problem is as I wrote above: this project is dead, deal with it. The logical successor is https://github.com/takari/io.takari.incrementalbuild

Comment (olamy) [reply src/main/java/org/apache/maven/shared/filtering/DefaultMavenFileFilter.java:109]: frankly looking at commit history m2e has more maintenance than https://github.com/takari/io.takari.incrementalbuild (last commit almost 2 yo ago).

Comment (cstamas) [reply src/main/java/org/apache/maven/shared/filtering/DefaultMavenFileFilter.java:109]: And what does it has to do with "liveliness" of 12 years dead project? Are you mixing up things maybe? sonatype/sisu-build-api is NOT m2e, so unsure what m2e "activity" has to do here. Also, you commented above how this API is "stable" (so stable that it did not need change for 12 years), and now you rant about measly 2 years?  :smile:

Comment (cstamas) [reply src/main/java/org/apache/maven/shared/filtering/DefaultMavenFileFilter.java:109]: If I get it correctly, what you say is "buildcontext prevails over takari/incrementalbuild as m2e has more commits"? :kangaroo: LOL, well, I am about to stop here.... :laughing:

Comment (olamy) [reply src/main/java/org/apache/maven/shared/filtering/DefaultMavenFileFilter.java:109]: > If I get it correctly, what you say is "buildcontext prevails over takari/incrementalbuild as m2e has more commits"? 🦘 LOL, well, I am about to stop here.... 😆

no sorry but you do not read correctly, the most important is the number of users impacted.
and here in m2e is a lot!
what is better it's not the question here. 
again I'm just saying the impacted users number is really huge!

Comment (cstamas) [reply src/main/java/org/apache/maven/shared/filtering/DefaultMavenFileFilter.java:109]: > frankly looking at commit history m2e has more maintenance than https://github.com/takari/io.takari.incrementalbuild (last commit almost 2 yo ago).

And who said this then? 

Anyway, lets bury the boomerang, sisu build api is raised from the dead

Comment (cstamas) [reply src/main/java/org/apache/maven/shared/filtering/DefaultMavenFileFilter.java:109]: ... and my 5 cents to whole this topic: I never saw a "plain" (explained later) Maven project that is fairly more complex than "hello world" and works OOTB in m2e. Work, as in no "tweaks", "m2e profile in POM" needed etc. OTOH, I did saw Maven project using takari lifecycle that WORK perfectly and OOTB in m2e. 

But alas, takari lifecycle use the "inferior" incrementalbuild and not buildcontext API. Oh, and it also replaces many things from Maven itself, and even ditches this maven-filtering and replaces it's with it's own to be functional :smile: 

("plain" as Maven build w/o any extension, while takari-lifecycle is an extension and replaces a LOT from maven, like core copmponets but also shared stuff like this maven-filtering and many plugins as well)

File: pom.xml [hunk 2]
```diff
@@ -141,21 +136,39 @@
...
-        <artifactId>plexus-component-metadata</artifactId>
-        <executions>
-          <execution>
-            <goals>
-              <goal>generate-metadata</goal>
-            </goals>
-          </execution>
-        </executions>
+        <groupId>org.eclipse.sisu</groupId>
+        <artifactId>sisu-maven-plugin</artifactId>
```
Comment (michael-o) [inline pom.xml:171]: Are you certain that this isn't in the parent already?

File: src/main/java/org/apache/maven/shared/filtering/DefaultMavenResourcesFiltering.java [hunk 3]
```diff
@@ -90,23 +87,25 @@ public boolean filteredFileExtension( String fileName, List<String> userNonFilte
...
+    }
+
     @Override
     public List<String> getDefaultNonFilteredFileExtensions()
     {
-        if ( this.defaultNonFilteredFileExtensions == null )
-        {
-            this.defaultNonFilteredFileExtensions = new ArrayList<>();
-        }
         return this.defaultNonFilteredFileExtensions;
```
Comment (michael-o) [inline src/main/java/org/apache/maven/shared/filtering/DefaultMavenResourcesFiltering.java:109]: This changes semantics, no?

**Review by michael-o — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 2 above)
Comment (cstamas) [reply pom.xml:171]: Is, in build/pluginMgmt, but i want it to execute

File: src/main/java/org/apache/maven/shared/filtering/DefaultMavenFileFilter.java [hunk 4]
```diff
@@ -19,32 +19,36 @@
...
+@Singleton
+@Named
 public class DefaultMavenFileFilter
     extends BaseFilter
     implements MavenFileFilter
 {
+    private final BuildContext buildContext;
 
-    @Requirement
-    private MavenReaderFilter readerFilter;
```
[the remark this answers is not in the document]
Comment (cstamas) [reply src/main/java/org/apache/maven/shared/filtering/DefaultMavenFileFilter.java:44]: Yup. Was unused member

File: src/main/java/org/apache/maven/shared/filtering/DefaultMavenResourcesFiltering.java (hunk 3 above)
Comment (cstamas) [reply src/main/java/org/apache/maven/shared/filtering/DefaultMavenResourcesFiltering.java:109]: Nope, member made final, no more "lazy init"

File: pom.xml (hunk 2 above)
Comment (michael-o) [reply pom.xml:171]: Makes sense!

File: src/main/java/org/apache/maven/shared/filtering/DefaultMavenResourcesFiltering.java (hunk 3 above)
Comment (michael-o) [reply src/main/java/org/apache/maven/shared/filtering/DefaultMavenResourcesFiltering.java:109]: Dammit, not noticed.

**Review by michael-o — APPROVED** [review]

**Comment by cstamas** [comment]
So, given the m2e team response (or better: lack of it), I think that maven-filtering is actually not in use in m2e (is being completely replaced), or, that buildcontext seems totally irrelevant @olamy

**Comment by fbricon** [comment]
Sorry for the late response (I'm not super active on the m2e front), but yeah the buildcontext API is very much used by m2e: https://github.com/eclipse-m2e/m2e-core/search?q=BuildContext
Basically `mvn resources:resources` is called on each file change, as part of Eclipse's incremental build. So if filtering where to be invoked on all resources instead of what's defined in the build context, we'd see severe performance degradation in Eclipse
